java/pattern/constraint/ValueConstraint.java

```diff
@@ -0,0 +1,663 @@
+
+                public ThingVar(ThingVariable<?> variable) {
+                    this.variable = variable;
+                    this.hash = Objects.hash(ThingVar.class, variable);
+                }
+
+                public ThingVariable<?> variable() {
+                    return variable;
+                }
+
+                @Override
+                public boolean isThingVar() {
+                    return true;
+                }
+
+                @Override
+                public ThingVar asThingVar() {
+                    return this;
+                }
+
+                @Override
+                protected void collectVariables(Set<BoundVariable> collector) {
+                    collector.add(variable);
+                }
+
+                @Override
+                public String toString() {
+                    return variable.reference().toString();
+                }
+
+                @Override
+                public boolean equals(Object o) {
+                    if (this == o) return true;
+                    if (o == null || getClass() != o.getClass()) return false;
+                    ThingVar that = (ThingVar) o;
+                    return this.variable.equals(that.variable);
+                }
+
+                @Override
+                public int hashCode() {
+                    return hash;
+                }
+            }
+
+            public static class ValVar extends Expression {
+                private final ValueVariable variable;
+                private final int hash;
+
+                public ValVar(ValueVariable variable) {
+                    this.variable = variable;
+                    this.hash = Objects.hash(ValVar.class, variable);
+                }
+
+                public ValueVariable variable() {
+                    return variable;
+                }
+
+                @Override
+                public boolean isValVar() {
+                    return true;
+                }
+
+                @Override
+                public ValVar asValVar() {
+                    return this;
+                }
+
+                @Override
+                protected void collectVariables(Set<BoundVariable> collector) {
+                    collector.add(variable);
+                }
+
+                @Override
+                public String toString() {
+                    return variable.reference().toString();
+                }
+
+                @Override
+                public boolean equals(Object o) {
+                    if (this == o) return true;
+                    if (o == null || getClass() != o.getClass()) return false;
+                    ValVar that = (ValVar) o;
+                    return this.variable.equals(that.variable);
+                }
+
+                @Override
+                public int hashCode() {
+                    return hash;
+                }
+            }
+
+            public abstract static class Constant<T> extends Expression {
```

Comment:
this should be called a `Value`, which also matches the grammar. Update the Expression builder too

Reply:
Changed the grammar to Constant, since that feels more correct.